# Assignment 1.1

In [2]:
import matplotlib as plt
import numpy as np
import pandas as pd
import sklearn
import sklearn.preprocessing as pre
from matplotlib.pyplot import plot
from sklearn.model_selection import train_test_split

## Task 1 Linear regression 

In [3]:
# 读出前五位
cpu = pd.read_csv('../../dataset/cpu_performance/machine.data',header=None)
cpu.columns = ['vendor', 'name', 'MYCT', 'MMIN', 'MMAX', 'CACH', 'CHMIN','CHMAX','PRP','ERP']
cpu.head()

vendor     name  MYCT  MMIN   MMAX  CACH  CHMIN  CHMAX  PRP  ERP
0  adviser    32/60   125   256   6000   256     16    128  198  199
1   amdahl   470v/7    29  8000  32000    32      8     32  269  253
2   amdahl  470v/7a    29  8000  32000    32      8     32  220  253
3   amdahl  470v/7b    29  8000  32000    32      8     32  172  253
4   amdahl  470v/7c    29  8000  16000    32      8     16  132  132

- Use the first 8 attribute to predict the 9th attribute, i,e, **published relative performance (PRP)**

- Build a **linear regression model** for this task, by using **MSE (mean square error)** loss function,
you are expected to get a **loss lower than 6000** on you test set

In [106]:
cpu_X, cpu_y = cpu.iloc[:, 0:8], cpu.iloc[:, -2]

In [5]:
cpu_y.shape

(209,)

- 查看cpu来源数量

In [6]:
cpu_vendor, cpu_name = cpu.iloc[:, 0], cpu.iloc[:, 1]

In [7]:
vendor_set = set(cpu_vendor)
print(vendor_set)
print(len(vendor_set))

{'adviser', 'magnuson', 'ibm', 'c.r.d', 'four-phase', 'ncr', 'microdata', 'bti', 'dec', 'amdahl', 'honeywell', 'perkin-elmer', 'apollo', 'cambex', 'ipl', 'hp', 'wang', 'cdc', 'prime', 'dg', 'gould', 'formation', 'burroughs', 'harris', 'siemens', 'sperry', 'sratus', 'basf', 'nas', 'nixdorf'}
30


- 产看cpu名字数量

In [8]:
name_set = set(cpu_name)
print(name_set)
print(len(name_set))

{'v8565:ii', 'dps:8/44', '3033:u', 'f4000/300ap', '7.541', '7.872-2', '4436', 'dn320', 'm80/32', '80/5', 'b2925', 'as/9000-dpc', '1636-1', '470v/7b', 'as/5000-n', 'b6900', 'universe:68/05', '3081:d', 'v8585:ii', '580-5860', '4331-11', '8130-b', '3210', 'as/5000', 'eclipse:mv/8000-ii', 'v8545:ii', '7.531', 'eclipse:mv/8000', '80/6', '4341-1', '8000', '1100/61-h1', '370/148', 'universe:68/37', '370/158-3', 'eclipse:c/350', 'as/5000-e', '3083:e', 'as/3000', 'b2900', '8890/30', 'vs-100', 'microvax-1', 'as/9060', 'dps:8/50', 'm80/44', 'b5920', 'v8595:ii', '50-250-ii', '80/3', '50-2250', '32/60', '1641-1', '470v/7', '600', '3083:b', '8130-a', 'cyber:170/835', '1100/81', 'as/8060', 'vs-90', '580-5840', 'omega:480-iii', 'dps:6/92', '4341-11', '7/68', 'b6925', '50-850-ii', '1100/83', '7.865-2', '38/5', 'vax:11/750', 'm80/30', '470v/b', '7.880-2', 'omega:480-ii', 'v8535:ii', '38/7', 'as/8040', '1100/82', 'dps:7/65', 'dn420', '3205', '7.870-2', '4381-1', '80/4', 'cyber:170/825', 'cyber:170/750', 

鉴于cpu的名称数量（209）与整个数据集cpu数量（209）相等，因此首先考虑了先不使用第二行信息进行模型构建，查看效果！

### 第一次实验：除了`name`字段进行模型构建 

In [38]:
# 取除了第二行name以外的数据用于构建
# cpu_X, cpu_y = cpu.iloc[:, 0:8], cpu.iloc[:, -2]
cpu_X_no_name = cpu.ix[:,[0,2,3,4,5,6,7]]

- 对于cpu供应商进行label构建

In [39]:
cpu_X_no_name.iloc[:, :].head()

vendor  MYCT  MMIN   MMAX  CACH  CHMIN  CHMAX
0  adviser   125   256   6000   256     16    128
1   amdahl    29  8000  32000    32      8     32
2   amdahl    29  8000  32000    32      8     32
3   amdahl    29  8000  32000    32      8     32
4   amdahl    29  8000  16000    32      8     16

In [40]:
cpu_X_no_name.iloc[:, 0] = pre.LabelEncoder().fit_transform(cpu_X_no_name.iloc[:, 0])

C:\Users\HYJ\Anaconda3\lib\site-packages\pandas\core\indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [41]:
cpu_X_no_name, cpu_y = np.array(cpu_X_no_name), np.array(cpu_y)

In [42]:
cpu_X_no_name.dtype

dtype('int64')

- Normalization & Zero-centered

In [84]:
XTrain1, XTest1, yTrain1, yTest1 = train_test_split(cpu_X_no_name, cpu_y, test_size=0.3, random_state=10)
featMean1, featStd1 = np.mean(XTrain1, axis=0), np.std(XTest1, axis=0)
XTrain1 = (XTrain1 - featMean1) / featStd1
XTest1 = (XTest1 - featMean1) / featStd1

In [85]:
print(XTrain1.shape)
print(featMean1.shape)

(146, 7)
(7,)


- Parameters Initialization 模型初始化

In [119]:
# 可以进行其他参数的初始化选择，如ones, rand等等
theta1 = np.zeros(XTrain1.shape[1]+1)

XTrain1 = np.concatenate((np.ones((XTrain1.shape[0], 1)), XTrain1), axis=1)
XTest1 = np.concatenate((np.ones((XTest1.shape[0], 1)), XTest1), axis=1)

In [120]:
initRate = 0.0001
learningRate = initRate

numEpoch = 1000
lossList = []
for epoch in range(numEpoch):
    # forward
    hyp = XTrain1@theta1
    MSE = 0.5/yTrain1.shape[0]*sum((hyp-yTrain1)**2)
    grad = (hyp-yTrain1)@XTrain1
    theta1 -= learningRate*grad
    if epoch % 50 == 0:
        print('Epoch', epoch, 'loss:', MSE)

Epoch 0 loss: 18069.8184932
Epoch 50 loss: 2064.07476429
Epoch 100 loss: 1688.64084103
Epoch 150 loss: 1588.42197973
Epoch 200 loss: 1557.79712726
Epoch 250 loss: 1547.48592015
Epoch 300 loss: 1543.73469017
Epoch 350 loss: 1542.26147655
Epoch 400 loss: 1541.62666498
Epoch 450 loss: 1541.31971457
Epoch 500 loss: 1541.15142598
Epoch 550 loss: 1541.04819128
Epoch 600 loss: 1540.97941399
Epoch 650 loss: 1540.93114513
Epoch 700 loss: 1540.89624896
Epoch 750 loss: 1540.87061455
Epoch 800 loss: 1540.85162621
Epoch 850 loss: 1540.83750049
Epoch 900 loss: 1540.82696914
Epoch 950 loss: 1540.81910882


In [97]:
# calculate the testing error of theta and compare it with algebraic solution
hypTest = XTest1@theta1
MSETest = 0.5/yTest1.shape[0]*sum((hypTest-yTest1)**2)
print("test error using gradient descent: ", MSETest)

test error using gradient descent:  2664.34395049


- ### 第二次实验：全部字段进行模型构建 

In [107]:
cpu_X.iloc[:, :].head()

vendor     name  MYCT  MMIN   MMAX  CACH  CHMIN  CHMAX
0  adviser    32/60   125   256   6000   256     16    128
1   amdahl   470v/7    29  8000  32000    32      8     32
2   amdahl  470v/7a    29  8000  32000    32      8     32
3   amdahl  470v/7b    29  8000  32000    32      8     32
4   amdahl  470v/7c    29  8000  16000    32      8     16

In [24]:
cpu_X.iloc[:, 0] = pre.LabelEncoder().fit_transform(cpu_X.iloc[:, 0])

In [108]:
cpu_X.iloc[:, 0] = pre.LabelEncoder().fit_transform(cpu_X.iloc[:, 0])
cpu_X.iloc[:, 1] = pre.LabelEncoder().fit_transform(cpu_X.iloc[:, 1])

In [109]:
cpu_X, cpu_y2 = np.array(cpu_X), np.array(cpu_y)

In [110]:
cpu_X.dtype

dtype('int64')

In [111]:
XTrain2, XTest2, yTrain2, yTest2 = train_test_split(cpu_X, cpu_y, test_size=0.3, random_state=10)
featMean2, featStd2 = np.mean(XTrain2, axis=0), np.std(XTrain2, axis=0)
XTrain2 = (XTrain2 - featMean2) / featStd2
XTest2 = (XTest2 - featMean2) / featStd2

In [113]:
print(XTrain2.shape)
print(featMean2.shape)

(146, 8)
(8,)


In [117]:
theta2 = np.zeros(XTrain2.shape[1]+1)

# Concatenate X with a new dimension for bias
XTrain2 = np.concatenate((np.ones((XTrain2.shape[0], 1)), XTrain2), axis=1)
XTest2 = np.concatenate((np.ones((XTest2.shape[0], 1)), XTest2), axis=1)

In [118]:
learningRate2 = 0.0001
# hyp = XTrain@theta
# MSE = 0.5/yTrain.shape[0]*sum((hyp-yTrain)**2)
# grad = (hyp-yTrain)@XTrain
# theta -= learningRate*grad

# train a linear model using SGD

numEpoch = 1000
lossList = []
for epoch in range(numEpoch):
    # forward
    hyp2 = XTrain2@theta2
    MSE2 = 0.5/yTrain2.shape[0]*sum((hyp2-yTrain2)**2)
    grad2 = (hyp2-yTrain2)@XTrain2
    theta2 -= learningRate2*grad2
    if epoch % 50 == 0:
        print('Epoch', epoch, 'loss:', MSE2)

Epoch 0 loss: 18069.8184932
Epoch 50 loss: 2305.99649681
Epoch 100 loss: 1692.24165346
Epoch 150 loss: 1588.86971857
Epoch 200 loss: 1555.07491639
Epoch 250 loss: 1541.36310387
Epoch 300 loss: 1535.10246863
Epoch 350 loss: 1531.96332712
Epoch 400 loss: 1530.26444664
Epoch 450 loss: 1529.28732991
Epoch 500 loss: 1528.69716733
Epoch 550 loss: 1528.32594944
Epoch 600 loss: 1528.08417539
Epoch 650 loss: 1527.92184776
Epoch 700 loss: 1527.80992968
Epoch 750 loss: 1527.73098086
Epoch 800 loss: 1527.67420001
Epoch 850 loss: 1527.63270128
Epoch 900 loss: 1527.60197249
Epoch 950 loss: 1527.57897919


In [121]:
hypTest2 = XTest2@theta2
MSETest2 = 0.5/yTest2.shape[0]*sum((hypTest2-yTest2)**2)
print("test error using gradient descent: ", MSETest2)

test error using gradient descent:  2631.60369625


### 结果

结果表明，`name`字段对于结果的预测有一定的帮助，但是过多的编码在遇到过多数据用于训练时，如name字段有209个数据则可能造成模型的爆炸，未来可以考虑先进性特征工程，对于`name`与`vendor`字段进行数据分析，优化模型